plugins/inputs/tacacs/tacacs.go

```diff
@@ -0,0 +1,150 @@
+	Servers         []string        `toml:"servers"`
+	Username        config.Secret   `toml:"username"`
+	Password        config.Secret   `toml:"password"`
+	Secret          config.Secret   `toml:"secret"`
+	RemAddr         string          `toml:"request_ip"`
+	ResponseTimeout config.Duration `toml:"response_timeout"`
+	Log             telegraf.Logger `toml:"-"`
+	clients         []tacplus.Client
+	testAuthStart   tacplus.AuthenStart
+}
+
+//go:embed sample.conf
+var sampleConfig string
+
+func (t *Tacacs) SampleConfig() string {
+	return sampleConfig
+}
+
+func (t *Tacacs) Init() error {
+	if len(t.Servers) == 0 {
+		t.Servers = []string{"127.0.0.1:49"}
+	}
+	if net.ParseIP(t.RemAddr) == nil {
+		return fmt.Errorf("Invalid ip address provided for request_ip: %s", t.RemAddr)
+	}
+
+	for index, serverValue := range t.Servers {
+		t.clients = append(t.clients, tacplus.Client{})
+		t.clients[index].Addr = serverValue
+		t.clients[index].ConnConfig = tacplus.ConnConfig{}
+	}
+
+	t.testAuthStart = tacplus.AuthenStart{
+		Action:        tacplus.AuthenActionLogin,
+		AuthenType:    tacplus.AuthenTypeASCII,
+		AuthenService: tacplus.AuthenServiceLogin,
+		PrivLvl:       1,
+		Port:          "heartbeat",
+		RemAddr:       t.RemAddr,
+	}
+
+	return nil
+}
+
+func (t *Tacacs) Gather(acc telegraf.Accumulator) error {
+	var wg sync.WaitGroup
+
+	for idx := range t.clients {
+		wg.Add(1)
+		go func(client *tacplus.Client) {
+			defer wg.Done()
+			acc.AddError(t.pollServer(acc, client))
+		}(&t.clients[idx])
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (t *Tacacs) pollServer(acc telegraf.Accumulator, client *tacplus.Client) error {
+	// Create the fields for this metric
+	tags := map[string]string{"source": client.Addr}
+	fields := make(map[string]interface{})
+
+	secret, err := t.Secret.Get()
+	if err != nil {
+		return fmt.Errorf("getting secret failed: %w", err)
+	}
+	defer config.ReleaseSecret(secret)
+
+	client.ConnConfig.Secret = secret
+
+	username, err := t.Username.Get()
+	if err != nil {
+		return fmt.Errorf("getting username failed: %w", err)
+	}
+	defer config.ReleaseSecret(username)
+
+	password, err := t.Password.Get()
+	if err != nil {
+		return fmt.Errorf("getting password failed: %w", err)
+	}
+	defer config.ReleaseSecret(password)
+
+	// send the start request, the reply should be AuthenStatusGetUser
+	ctx := context.Background()
+	if t.ResponseTimeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(t.ResponseTimeout))
+		defer cancel()
+	}
+
+	startTime := time.Now()
+	reply, session, err := client.SendAuthenStart(ctx, &t.testAuthStart)
+	if err != nil {
+		return fmt.Errorf("error on new tacacs authentication start request to %s : %w", client.Addr, err)
+	}
+	defer session.Close()
+	// Check the returned status
+	if reply.Status != tacplus.AuthenStatusGetUser {
+		return fmt.Errorf("error on new tacacs authentication start request to %s : Unexpected response code %d", client.Addr, reply.Status)
+	}
```

Comment:
I don't remember what we did on radius, but it might make sense to have this in the metric to make it discoverable to monitoring....

Reply:
in radius we set responsetime to maximum timeout value if its not success and use response code in tag, Ill do that, although its harder to debug for users based on that info due to the multi-stage auth rather than single stage as in radius. (He wont have info from which stage the response code is - that is currently provided by the hard error message). Optionally we could do custom status from this set: ["success", "error on new tacacs authentication start request to %s : Unexpected response code %d", "error on first tacacs authentication continue username request to %s : Unexpected response code %d", "error on second tacacs authentication continue password request to %s : Unexpected response code %d"] which would provide more immediate info. Let me know what you think. For now commiting it with the raw values.